### DISCUSSION TITLE
200 response with an embedded 401 for user info

### DISCUSSION THREAD

**Discussion by adenix** [OpenIDC/mod_auth_openidc#1077] [General]
We have `mod_auth_openidc` implemented at scale with a shared redis cache. Everything has been working great until our IdP vendor updated their userinfo endpoint to return `401` on expired `access_token` instead of `403`. We now are seeing odd behavior at end of session. The `access_token` is getting refreshed but the userinfo is returned as an embedded `401`. My theory is that the new status code is causing the error response to be cached.

```
LoadModule authn_core_module modules/mod_authn_core.so
LoadModule authz_user_module modules/mod_authz_user.so
LoadModule auth_openidc_module modules/mod_auth_openidc.so
OIDCDefaultURL https://insight-meechum.prod.netflix.net
OIDCInfoHook access_token access_token_expires refresh_token userinfo
OIDCProviderMetadataURL https://meechum.netflix.com/.well-known/openid-configuration
OIDCClientID insight_meechum
OIDCClientSecret ${meechum_secret}
OIDCRedirectURI https://insight-meechum.prod.netflix.net/meechum
OIDCCryptoPassphrase <REDACTED>
OIDCScope <REDACTED>
OIDCPassClaimsAs headers
OIDCSessionInactivityTimeout 28800
OIDCSessionMaxDuration 28800
OIDCPassRefreshToken Off
OIDCStripCookies mod_auth_openidc_session mod_auth_openidc_session_chunks mod_auth_openidc_session_0 mod_auth_openidc_session_1
OIDCUnAuthAction auth
OIDCScrubRequestHeaders On
OIDCCacheType redis
OIDCOAuthVerifyJwksUri https://meechum.netflix.com/ext/oauth/nflxJwTDefault/jwks
OIDCUserInfoRefreshInterval 900
OIDCRedisCacheServer <REDACTED>
OIDCHTMLErrorTemplate /apps/meechum/error.html
```

![Screenshot 2023-07-10 at 3 26 51 PM](https://github.com/OpenIDC/mod_auth_openidc/assets/8764760/d11b7503-7dea-4fee-9d45-e623e5a32299)

**Comment by zandbelt** [comment]
what version of mod_auth_openidc are you on?

**Reply by adenix** [reply]
Looks like `2.3.8` for this specific service. Running an older OS `Ubuntu 18.04.6 LTS(bionic)`.

**Comment by zandbelt** [comment]
upgrading to a recent release should solve this issue then: it will check for a "sub" claim

**Reply by adenix** [reply]
Do you know what version is the threshold for this issue? We upgrade to `2.4.11-1` in our Ubuntu Jammy OS.

**Comment by zandbelt** [comment]
seems to be 2.4.5

**Comment by adenix** [comment]
No longer seeing the 401 embedded response on refresh but now the userinfo is being completly omitted on refresh.

<img width="912" alt="Screenshot 2023-07-11 at 1 59 00 PM" src="https://github.com/OpenIDC/mod_auth_openidc/assets/8764760/f0cacc3a-173c-448e-8a2e-530f8e9cfb8d">

**Reply by zandbelt** [reply]
isn't that what one would expect after a 401 is returned?

**Reply by adenix** [reply]
No. We expect the userinfo endpoint to have been called again with the valid `access_token` after it was refreshed. Here the refreshed `access_token` is being displayed but the userinfo is being omitted.

**Reply by zandbelt** [reply]
please check the server log file with `LogLevel` set to `auth_openidc:debug` to see what is going on

**Reply by adenix** [reply]
Looks like its successfully refreshing the AT, successfully getting the user info with a `sub` claim, then trying refresh the AT again with the old RT, and getting an error as it should.

[auth_openidc.error.log](https://github.com/OpenIDC/mod_auth_openidc/files/12022419/auth_openidc.error.log)

**Reply by zandbelt** [reply]
that appears to be a bug indeed; would you be able to test a fix from https://mod-auth-openidc.org/downloads/libapache2-mod-auth-openidc_2.4.14.3rc2-1.bionic_amd64.deb and https://www.mod-auth-openidc.org/downloads/libapache2-mod-auth-openidc_2.4.14.3rc2-1.jammy_amd64.deb?

**Reply by adenix** [reply]
Testing, I'll get back with the results 🙇

**Reply by adenix** [reply]
Sorry for the delay. Initial testing didn't go as expected. I'm planning to do more testing today.

**Reply by adenix** [reply]
@zandbelt testing finished and the RC is behaving as expected

**Comment by adenix** [comment]
@zandbelt Do you have an idea of when this will make it into a full release? I'm just wanting to set expectations internally.

**Reply by zandbelt** [reply]
the release schedule is driven by our commercial support customers (hint ;-))

**Reply by adenix** [reply]
@zandbelt That makes sense for a feature request, but this is a bug fix. Do you have any idea when this will be released?
